### PR TITLE
Enumerable | Rspec Testing

### PR DESCRIPTION
**Pull request by uzairali19**
# Rspec Testing

## Advanced Building Blocks - Enumerables

The project is based on reproducing the enumerable methods to their usable states. The functions would have the prefix `#my_` attached to them, eg: the `#each` function would be `#my_each`. These functions would have tests related to their scopes. The tests can be accessed by running the `bundle install` and `rake` commands on the terminal.

Functions to test :

- `#my_each`
- `#my_each_with_index`
- `#my_select`
- `#my_all?`
- `#my_any?`
- `#my_none?`
- `#my_count`
- `#my_map`
- `#my_inject`

## Authors

👤 **Uzair**

- LinkedIn: [LinkedIn](https://www.linkedin.com/in/uzair-ali-964187166/)
- GitHub: [@githubhandle](https://github.com/uzairali19)

### CODE REVIEW

File: spec/enumerable_spec.rb [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+  array = [1, 2, 3, 4, 5, 6, 7, 8]
+  string_array = %w[uzair emanuel dan vic Oni]
+  hash = { one: 1, two: 2, three: 3 }
+
+  describe '#my_each' do
+    context 'If no block given' do
+      it 'returns enum' do
+        expect(array.my_each).to be_an Enumerator
+      end
+    end
```
Comment (elisha2kyakpo1) [inline spec/enumerable_spec.rb:13]: - [x] This is nice, but you should also test for negative scenarios.

File: spec/enumerable_spec.rb [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+        sum = :+.to_proc
+        new_array = array.my_inject(&sum)
+        expect(new_array).to eq(36)
+      end
+    end
+
+    context 'If a block is given and argument is given' do
+      it 'returns the largest word from string_array' do
+        new_array = string_array.my_inject([]) { |memo, word| memo.size > word.size ? memo : word }
+        expect(new_array).to eq('emanuel')
```
Comment (elisha2kyakpo1) [inline spec/enumerable_spec.rb]: - [x] Test for negative scenarios.

**Review by elisha2kyakpo1 — CHANGES_REQUESTED** [review]
# Status: Changes required👈
Hi, @uzairali19 you have done a really good job, you are almost there.

**To highlight:**

- All the tests are well written and are all passing.

**Changes to make:**

- [x] Add negative tests as well, this is good because it will show how good the methods are. For example, this test below is the opposite of the original.
```
    context 'If a block is given and any conditions donot pass' do
      it 'returns false' do
        self_array = array.any? do |num|
          num < 10
        end
        expect(self_array).to true
      end
    end
```
- [x] I have left a similar comment below.

<hr>

### Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

File: spec/enumerable_spec.rb (hunk 1 above)
Comment (uzairali19) [reply spec/enumerable_spec.rb:13]: Hello,

I am very confused. Can you please give an example where:

```
context "If no block given" do
     it "returns enum" do
        expect(array.my_each).to be_an Enumerator
      end
    end
```

In this case, the negative scenario for each statement is that if there is nothing you are passing on the function eg: `#my_each`, It will always return an enumerator. As per the Ruby API documentation.

For each of the functions in my enumerable module, I have tried to test them for arrays, string_arrays, and hashes.
Some of the functions must have a negative scenario, I understand it for example I can think of some negative scenarios for `#my_any?`, `#my_all?` and `#my_none?` and I will submit after implementing them. 

As for the `#my_each` function, I cannot think of any other negative scenario where the `if a block is not given` `#my_each` function should perform differently.

Waiting for your response!

Happy Coding!

Thanks,
Uzair Ali

Comment (elisha2kyakpo1) [reply spec/enumerable_spec.rb:13]: Hell @uzairali19 hope you are fine, I get your point. I understand that my_each(each) function should iterate through an array and will return each array element so that we can work or perform a task, so in this case, something like this can work.
 describe '#my_each' do
    it 'Performs a block of code if given and returns the passed enumerable' do
      expect(my_array.my_each { |x| 5 * x }).to eq(my_array)
    end
    it 'Performs a block of code if given and returns the passed enumerable' do
      expect(my_array.my_each { |x| 5 * x }).to_not eq([1, 2])
    end
  end

Comment (uzairali19) [reply spec/enumerable_spec.rb:13]: Hello,

Thank you for the clarification.

Regards,
Uzair Ali

**Review by BrianSammit — APPROVED** [review]
STATUS:APPROVED :green_circle: :+1: 
--------------
Hello @uzairali19, great job working on the required changes :muscle: Your test are working properly and you make a great number of tests. :100:
-----------
Now you can merge this branch :clap: :1st_place_medal: 
HAPPY CODING!! :four_leaf_clover: :smile: 

well-1
